index.d.ts

```diff
@@ -796,6 +797,20 @@ declare namespace plugins {
      * @default true
      */
     signature?: boolean;
+
+    /**
+     * An object of optional callbacks to be executed during the respective
+     * phase of a GraphQL operation. Undefined callbacks default to a noop
+     * function.
+     * 
+     * @defaul {}
+     */
+    hooks?: {
+      parse?: (span: Span, args: { source: string }) => void;
+      validate?: (span: Span, args: { source: string }) => void;
+      execute?: (span: Span, args: ExecutionArgs, res: ExecutionResult) => void;
```

Comment:
Should it be the `ExecutionContext` instead? It's closer to what is currently passed to other hooks.

Reply:
I've chosen `ExecutionArgs` because it has a better match with what is returned by the `normalizeArgs` function.

```ts
// https://github.com/graphql/graphql-js/blob/157edcb4f7dfe9cb2acd3d82831430d983bbfbcf/src/execution/execute.d.ts#L29
export interface ExecutionContext {
  schema: GraphQLSchema;
  fragments: { [key: string]: FragmentDefinitionNode }; // not returned by normalizeArgs
  rootValue: any;
  contextValue: any;
  operation: OperationDefinitionNode; // not returned by normalizeArgs (it's operationName instead)
  variableValues: { [key: string]: any };
  fieldResolver: GraphQLFieldResolver<any, any>;
  errors: Array<GraphQLError>; // not returned by normalizeArgs
}

// https://github.com/graphql/graphql-js/blob/157edcb4f7dfe9cb2acd3d82831430d983bbfbcf/src/execution/execute.d.ts#L51
export type ExecutionArgs = {
  schema: GraphQLSchema;
  document: DocumentNode;
  rootValue?: any;
  contextValue?: any;
  variableValues?: Maybe<{ [key: string]: any }>;
  operationName?: Maybe<string>;
  fieldResolver?: Maybe<GraphQLFieldResolver<any, any>>;
  typeResolver?: Maybe<GraphQLTypeResolver<any, any>>; // not returned by normalizeArgs but it's an optional property at least
};

// https://github.com/DataDog/dd-trace-js/blob/master/packages/datadog-plugin-graphql/src/index.js#L211-L219
function normalizeArgs (args) {
  if (args.length === 1) {
    return args[0]
  }

  return {
    schema: args[0],
    document: args[1],
    rootValue: args[2],
    contextValue: args[3],
    variableValues: args[4],
    operationName: args[5],
    fieldResolver: args[6]
  }
};
```

Reply:
Correct, but I'm still not sure it's the right approach.

If I compare this to HTTP for example, it would look like this:

```js
const req = http.get(options, (res) => {})
```

* `ExecutionArgs` is equivalent to `options`. It's provided only once by the user when the operation is executed.
* `ExecutionContext` is equivalent to `req`. It represents the actual request and is available throughout the library.
* `ExecutionResult` is equivalent to `res`. It represents the response and is available at the end of the operation.

In that sense, I'd lean toward `ExecutionContext` unless there is a strong argument for `ExecutionArgs`.

Reply:
When I'm looking at the definition of the two types, I don't see a huge conceptual difference. They both have 8 properties, out of which 3 differ (`ExecutionArgs` -> `ExecutionContext`):

1. `document` -> `fragments`,
2. `operationName` -> `operation`,
3. `typeResolver` -> `errors`.

In the first case, `ExecutionArgs` provides more information because fragments can be derived from the document using the following code:
```js
const fragments = document && document.definitions.reduce((acc, d) => {
  if (d.kind !== 'FragmentDefinition') return acc
  return {
    ...acc,
    [d.name.value]: d
  }
}, {})
```

In the second case, `ExecutionContext` provides more information because `operationName` can be derived from `operation`.

In the last case, `ExecutionArgs` provides more information because the typeresolver is missing from `ExecutionContext` and `errors` can be derived from `ExecutionResult`.

It may be desirable to provide as much information as possible to the hook. That could be achieved with `ExecutionArgs` where `operationName` would be replaced with `operation`, unless I'm missing something 🙂

Reply:
Appreciate all your hard work and detailed analysis on this topic, @vecerek and @rochdev!

Disclaimer: I am not intimately familiar with the two types in question. But I hope my perspective as an outsider will be of some value. While I appreciate the aim of providing as much information as possible to the hook, doing so comes at the expense of some conceptual clarity. As @rochdev points out, each of the types serves a particular purpose and has a particular meaning with well-established patterns around where and how they are intended to be used.

Therefore, I would strongly encourage the use of `ExecutionContext`.

Reply:
@vecerek I still think `ExecutionContext` makes the most sense for relevance and availability, even if maybe not for the overall raw data. It's also the safest option so that if we move the instrumentation deeper into the `grapghql` module we still have access to the `ExecutionContext` since it's what's available within the framework.

Again, all of the above also applies to HTTP since some information from `options` maybe not be available in `req`. I think this is an acceptable limitation since usually the information is still available but in a different form, for example a URL string vs a URL object, or a document vs fragments.

Of course it's possible that I'm completely missing some use cases, but at that point I think we could always add the `ExecutionArgs` as an additional parameter. The same could be said for HTTP if `options` is ever needed in a hook.

Reply:
Let's do that then 🙂Addressed in this commit:  https://github.com/DataDog/dd-trace-js/pull/909/commits/29fadaf2f23df64418cb50c53dfd43889593f729